### ISSUE TITLE
Mailgun Adapter: ArgumentError :erlang.bit_size(nil)

### ISSUE THREAD

**Issue by monicao** [swoosh/swoosh#76]
Using `swoosh: 0.3.0` and `phoenix_swoosh: 0.1.1`.

I am pretty sure this error is being raised because there is something wrong with my config. It is hard to figure out exactly what is missing from the error message.

```
> MyApp.Emails.forgot_password(%{first_name: "Test", last_name: "User", email: "myemail@gmail.com"}, "test") |> MyApp.Mailer.deliver
** (ArgumentError) argument error
             :erlang.bit_size(nil)
    (swoosh) lib/swoosh/adapters/mailgun.ex:33: Swoosh.Adapters.Mailgun.deliver/2
```

Quick link to the line in question: https://github.com/swoosh/swoosh/blob/v0.3.0/lib/swoosh/adapters/mailgun.ex#L33

Here is the adapter configuration.

```
> Application.get_env(:myapp, MyApp.Mailer)
[adapter: Swoosh.Adapters.Mailgun,
 api_key: "key-a1a2a2s22d2d2d2d2d2d2",
 domain: "mailgun.myapp.com"]
```

And the raw `sys.config` entry just in case this is an issue with strings vs binaries.

```
[
% ... other config stuff here
{myapp, [
  {'Elixir.MyApp.Mailer',[
    {adapter,'Elixir.Swoosh.Adapters.Mailgun'},
    {api_key,<<"key-a1a2a2s22d2d2d2d2d2d2">>},
    {domain,<<"mailgun.myapp.com">>}]}
]},
]
.
```

Any ideas why this error is happening? 

The code works in development where the swoosh config is defined in `config/dev.exs` instead of the `sys.config` file (exrm).

**Comment by barisbalic** [comment]
Hi @monicao thanks for raising this issue.  From what you have said, that sounds correct, although from your example I can't see what's wrong and I can reproduce it.  Would you be up for recreating the issue in a public repo?

**Comment by monicao** [comment]
Sure, I can give that a shot tomorrow.

**Comment by barisbalic** [comment]
That is hugely appreciated @monicao

**Comment by monicao** [comment]
I realized that in order to reproduce the issue I would have to set up the deployment environment (digital ocean droplet, install erlang, etc.). That will be very time consuming and there is no guarantee that it will work.

The email is being sent successfully in development, it's just not sending in production.

I'd be happy to do a quick screen sharing session to debug this on the server I already have. We can set that up over email.

**Comment by justinbkay** [comment]
I'm seeing this same error, in dev and production when trying to use the mailgun adapter on phoenix 1.1.3.

`
** (exit) an exception was raised:
    ** (ArgumentError) argument error
        :erlang.bit_size(nil)
        (swoosh) lib/swoosh/adapters/mailgun.ex:33: Swoosh.Adapters.Mailgun.deliver/2
        (nwrecc_org) web/controllers/about_controller.ex:43: NwreccOrg.AboutController.send_minisite_mail/2
`

**Comment by stevedomin** [comment]
Hi @monicao, I've done some digging on that issue and it looks like it's caused by the `domain` config being `nil`.

I'm trying to build a release now to check if I can reproduce the bug in that case.

**Comment by justinbkay** [comment]
I saw that error as well, it was because of how my config was being done.
Ended up just putting the key/domain in the same spot as my production
database info.  Was trying to load via App.get_env

On Fri, May 13, 2016 at 3:52 PM, Steve Domin <notifications@github.com>
wrote:

> Hi @monicao <https://github.com/monicao>, I've done some digging on that
> issue and it looks like it's caused by the domain config being nil.
>
> I'm trying to build a release now to check if I can reproduce the bug in
> that case.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/swoosh/swoosh/issues/76#issuecomment-219168230>
>



-- 
Sites:
http://jbkayconstruction.com
https://twitter.com/justinbkay
http://justinbkay.tumblr.com

It always takes longer than you expect, even when you take into account
Hofstadter’s Law.
- Hofstadter’s Law

**Comment by stevedomin** [comment]
You can load config via environment variable by using a `{:system, "ENV_VAR"}` tuple. It's described in the `Swoosh.Mailer` docs (https://hexdocs.pm/swoosh/Swoosh.Mailer.html).

I'm sorry this is hard to find, you need to make it clearer that this is possible.

**Comment by barisbalic** [comment]
Haha, I think @stevedomin meant. "we" need to make it clearer, not you!

**Comment by stevedomin** [comment]
Yep, definitely, sorry :)

**Comment by stevedomin** [comment]
@monicao I've tried extensively to reproduce that issue using releases but I haven't managed to.

I've put together an example application here: https://github.com/swoosh/phoenix_swoosh_example

Can you try using environment variables like it's done in this application and let us know if you are still seeing the same error. That would be very helpful, thanks!

Also, it would be great to see your code, it doesn't need to be on a server, just getting access to a repo would be enough. Let us know if that's possible, otherwise I'll drop you an email and we'll schedule that screen sharing session.

Thanks again, and sorry everything is not working as expected.

**Comment by monicao** [comment]
@stevedomin I am working on closed source app, so unfortunately I can't share the code directly.

I will try using environment variables to see if that changes things.

**Comment by monicao** [comment]
I ended up forking swoosh and inspecting the `config` variable that was passed into `deliver`.

The issue was that both the `:domain` and the `:api_key` were missing from config. It would be helpful if the adapter raised an exception if either of these two config values are missing.

I could do a quick PR for this if you think it's worth it.

**Comment by stevedomin** [comment]
Out of interest, do you know why they were missing?

**Comment by monicao** [comment]
I am using edeliver to deploy and I added the mailgun settings in my `sys.config` that was copied over using the [steps described here](https://github.com/boldpoker/edeliver/wiki/Use-per-host-configuration). It turns out I also needed to add the entries in my `production.secret.exs` file [as described here](https://github.com/boldpoker/edeliver/wiki/Embed-Secrets---Credentials-into-the-Release).

I didn't have time to investigate why this is happening, but with the duplicated config settings this works now. Might have something to do with the fact that one of these config files is read at compile time and one of them is available at runtime.
